pilot/pkg/bootstrap/istio_ca.go

```diff
@@ -286,6 +289,181 @@ func (s *Server) loadRemoteCACerts(caOpts *caOptions, dir string) error {
 	return nil
 }
 
+// isValidCACertsFile As we are watching entire directory, but interested
+// in only 'ca-key.pem', 'ca-cert.pem', 'root-cert.pem' and 'cert-chain.pem'.
+// Events on other files are ignored.
+func isValidCACertsFile(path string) (string, bool) {
+	_, file := filepath.Split(path)
+
+	for _, name := range []string{"ca-cert.pem", "ca-key.pem", "root-cert.pem", "cert-chain.pem"} {
+		if file == name {
+			return file, true
+		}
+	}
+
+	return "", false
+}
+
+// isValidCACertsEvent We are interested in only Create, Write
+// and Delete events. Other event types are ignored at the moment.
+func isValidCACertsEvent(event fsnotify.Event) uint8 {
+	if event.Op&fsnotify.Create == fsnotify.Create {
+		return create
+	} else if event.Op&fsnotify.Write == fsnotify.Write {
+		return write
+	} else if event.Op&fsnotify.Remove == fsnotify.Remove {
+		return remove
+	}
+
+	return invalid
+}
+
+// updateCACertsMap updates the map with cacerts files with
+// particular events only.
+func updateCACertsMap(s *Server, event fsnotify.Event) bool {
+	file, ok := isValidCACertsFile(event.Name)
+	if !ok {
+		return false
+	}
+
+	op := isValidCACertsEvent(event)
+	if op == invalid {
+		return false
+	}
+
+	s.cacertsMutex.Lock()
+	s.cacertsMap[file] = op
+	s.cacertsMutex.Unlock()
+	return true
+}
+
+func areCACertsFilesCreated(s *Server) bool {
+	if s.cacertsMap["ca-cert.pem"] == create &&
+		s.cacertsMap["ca-key.pem"] == create &&
+		s.cacertsMap["root-cert.pem"] == create &&
+		s.cacertsMap["cert-chain.pem"] == create {
+		return true
+	}
+
+	return false
+}
+
+func areCACertsFilesModified(s *Server) bool {
+	if s.cacertsMap["ca-cert.pem"] == write &&
+		s.cacertsMap["ca-key.pem"] == write &&
+		s.cacertsMap["root-cert.pem"] == write &&
+		s.cacertsMap["cert-chain.pem"] == write {
+		return true
+	}
+
+	return false
+}
+
+// handleEvent handles the events on cacerts related files.
+// If create/write(modified) event occurs, then it verifies that
+// newly introduced cacerts are intermediate CA which is generated
+// from cuurent root-cert.pem. Then it updates and keycertbundle
+// and generates new dns certs.
+// TODO(rveerama1): Add support for new ROOT-CA rotation also.
+func handleEvent(s *Server, event fsnotify.Event) {
+	ok := updateCACertsMap(s, event)
+	if !ok {
+		return
+	}
+
+	if areCACertsFilesCreated(s) || areCACertsFilesModified(s) {
```

Comment:
what if some of them have been created and others have been modified?

Reply:
Is there any way to do like that (I mean delete few and create those deleted ones and modify others) ? If yes, what are valid scenarios like that ( I mean what files are eligible to delete/create and what are for modifiable) ?

Reply:
You cannot write 4 files atomically, so wouldn't this happen in all cases for some period of time?

Reply:
sorry, did not understood clearly. Can you elaborate or give some suggestions to improvise it?